### PR TITLE
Library: Add ReadAsExpression and ReadAsList

### PR DESCRIPTION
**Pull request by markuspf**
This was suggested by Matan Ziv-Av on the GAP mailing list:

https://mail.gap-system.org/pipermail/gap/2015-February/000008.html

### CODE REVIEW

File: lib/files.gd [hunk 1]
```diff
@@ -316,6 +316,63 @@ DeclareOperation( "ReadTest", [ IsString ] );
...
+#############################################################################
+##
+#O  ReadAsExpression( <name-file> ) . . . . . . . . read a file as expression
+##
+##  <#GAPDoc Label="ReadAsExpression">
+##  <ManSection>
+##  <Oper Name="ReadAsExpression" Arg='name-file'/>
+##
+##  <Description>
+##  reads the file with filename <A>name-file</A> as a GAP expression.
```
Comment (frankluebeck) [inline lib/files.gd]: I would not use 'name-file' here because for GAP this means "name minus file", how about 'fname' instead?

**Comment by ChrisJefferson** [comment]
I like 'ReadAsExpression', but I don't really understand 'ReadAsList'. Is it that common for people to store a list with commas, but without [ and ] ?

File: lib/files.gd [hunk 2]
```diff
@@ -316,6 +317,63 @@ DeclareOperation( "ReadTest", [ IsString ] );
...
+##  <E>Example</E>
+##  <P/>
+##  Suppose that the file <F>/tmp/example.g</F> contains the following
+##  <P/>
+##  <Log><![CDATA[
+##  (1,2,3)
+##  ]]></Log>
+##  <P/>
+##  <Log><![CDATA[
+##  gap> a := ReadAsExpression("/tmp/example.g")();
```
Comment (ChrisJefferson) [inline lib/files.gd]: Where is there an extra () on the end of this line? (Also on end of 'ReadAsList' example)?

Comment (markuspf) [inline lib/files.gd]: Probably because I didn't clean up properly, I'll correct this.

**Comment by markuspf** [comment]
I don't know whether this is "common", I only implemented a suggestion by a user. I suppose if you have a list of generators of some description that might come without []. That said, we should have a good look at exchange formats or file parsers anyway.

I should probably also add some tests for these functions before this gets merged.

**Comment by alex-konovalov** [comment]
Replying to Chris, I think that the documentation of `ReadAsList` should be more precise about the format of the data. I am imagining as one of the common cases printing list to a file, and the proposed solution for `ReadAsList` allows to avoid editing the file with initial assignment and trailing semicolon. One should only to remember that in this case the list from the file will be in the 1st position of the list returned by `ReadAsList`. An ability to read also comma separated data comes as a nice add-on.

**Comment by ChrisJefferson** [comment]
Alex: Given a file containing `[1,2,3]` then ` a := ReadAsExpression("file")` will assign `a` the value `[1,2,3]`. If your file already contains a GAP list, then you *don't* want `ReadAsList`.

`ReadAsList` lets you have a file containing `1,2,3` and then `a := ReadAsList("file");` will put the `[` and `]` on the start and end of the file to turn it into a valid GAP list. This could be useful for some people I imagine, but perhaps the name is misleading...

I certainly wouldn't want this argument to stop `ReadAsExpression`, which is a really useful function.

**Comment by alex-konovalov** [comment]
I'd suggest to add only `ReadAsExpression` and not the other one. Of course, if there are genuine cases of  `ReadAsList` solving someone's problems, then I wouldn't mind adding it too. 

By the way, IIRC for large data files reading them with `Read` will be more efficient, because if will add new list entries to the list while it parses the file, without creating a string in the middle. If true, this should be probably reflected in the documentation.

**Comment by fingolfin** [comment]
I agree with Alex, simply because anything we add now, we may need to support indefinitely; and it adds to the overall complexity of GAP. I am fine with presenting such functions on a page with "useful GAP tidbits", but are they really useful for a general audience? I.e. is there at least one other person who would use these functions, besides the one originally asking about them?

Overall, I prefer add or improving existing functionality such that that writing things like `ReadAsList` becomes easy (ideally: "trivial") to do for end users. In this case, the code Markus' wrote is already a pretty thin wrapper around EvalString. 

My alternate implementations are even shorted (though they provide less nice error messages if the file is not found):
```
ReadAsExpression := name -> EvalString(StringFile(name));
ReadAsList := name  -> EvalString(Concatenation("[",StringFile(name),"]"));
```
Do we really need to add such things to the GAP library? I am doubtful...

On a tangential line of thought, though, consider that this is the current implementation of EvalString:
```
_EVALSTRINGTMP := 0;
InstallGlobalFunction("EvalString", function( s )
  local a, f, res;
  a := "_EVALSTRINGTMP:=";
  Append(a, s);
  Add(a, ';');
  Unbind(_EVALSTRINGTMP);
  f := InputTextString(a);
  Read(f);
  if not IsBound(_EVALSTRINGTMP) then
    Error("Could not evaluate string.\n");
  fi;
  res := _EVALSTRINGTMP;
  Unbind(_EVALSTRINGTMP);
  return res;
end);
Unbind(_EVALSTRINGTMP);
```

Now, it is indeed annoying to read a whole file into a string, muck with that string a bit, then turn it into a stream and pass that to `Read`. For some of my own code, I've been doing similar tricks. I am sure we can do better. Though this is certainly not something that needs or should be done as part of this PR -- just saying that there is IMHO a general need to improve the situation.

One possible way out would be to create a new type of streams: "union" or "join" streams. This input stream object type would simply wrap a set of streams S1, S2, S3; when reading from such a union stream, it would first read from the first stream, until end-of-stream; then from the second stream, etc.

Now, just create a "union stream" from one or more InputTextString streams plus one InputTextFile, and the overhead has been avoided.

**Comment by fingolfin** [comment]
BTW, note that my implementation in the end uses the kernel function `READ_STRING_FILE` to read the input file, while Markus' implementation boils down to `READ_ALL_FILE`. I don't think that this matters much for the argument, but found it curious and thought I should point it out.

**Comment by markuspf** [comment]
It seems a bit silly that everyone invents their own one-liners for those "trivial" tasks though. In that sense creating a package that contains "common utilities" becomes a viable option.

In particular with Max' comment about "having to carry things around indefinitely" in mind, I would even go and not *add* things to the library but rather *remove* things from the library and put them into packages?

I think we should be *much* more free with respect to adding functionality and deciding that things did not work out sometimes, because otherwise we end up discussing merits of even these tiny functions indefinitely.
And finally, rather inappropriately here (we should do these discussions about "general things" on the mailing list or in an appropriate issue...), we should discuss what we'll understand under "the Library" for the future, and what we think should go into packages.

**Comment by fingolfin** [comment]
Markus, there are millions of one-lines which are useful to somebody -- are you suggesting we add all of them? Which other language does that?

And I think you are a bit cavalier about removing things -- I know that acting this way is fashionable these days, and it makes sense when e.g. developing for the web, when things move fast. But in mathematics, programs are often used for decades, and their authors get quickly (and IMHO, rightfully) upset if those programs break when we remove things or (even worse) change the behaviour of something that's been around for ages.

Thus, while for my own projects and also for more "agile" projects, I am very much in favor of being liberal in adding and also removing things, I don't think this is a good when applied to any documented functionality in GAP (and, in fact, even when it comes to undocumented but nevertheless widely used functionality).

I am happy to continue this discussion on the mailing list, too, though I think most "old timers" will agree with me there (but they should speak for themselves, of course :).

**Comment by markuspf** [comment]
I updated the pull request to just add ReadAsExpression. The further discussion that happened here is material for an issue (UnionStreams) and the discussion I pulled out of here onto the gap mailing list.

**Comment by MatanZ** [comment]
My use case for ReadAsList is something like:
for i in [1..32] do AppendTo("hg",HaarGraphs(PSmallGroups(i)),","); od;
This program can take some time to run, but the file "hg" with partial results can always be read into another GAP session.

**Comment by frankluebeck** [comment]
If I was asked to decide, I would reject both suggested utilities (ReadAsList and ReadAsExpression)
because they are really easy to implement and because I doubt that they are of sufficiently general interest. 

It is more useful for GAP users to learn about StringFile and EvalString which should enable them to write the utilities discussed here and many other similar ones easily. Even finding out that such a thing as ReadAsExpression exists and studying its documentation is more complicated than writing it.

A side remark: Reading a file into a string with StringFile and then reading this string causes very little overhead compared to reading the file directly (even if the content of the file is  very easy to parse, say a list of integers).

**Comment by markuspf** [comment]
Bikeshed prevention.

**Pull request closed without merging**